File: src/androidTest/java/com/better/alarm/test/ListTest.java
package com.better.alarm.test;

import android.content.Intent;
import android.support.test.espresso.FailureHandler;
import android.support.test.rule.ActivityTestRule;
import android.support.test.runner.AndroidJUnit4;
import android.test.suitebuilder.annotation.LargeTest;
import android.view.View;

import com.better.alarm.R;
import com.better.alarm.configuration.AlarmApplication;
import com.better.alarm.interfaces.PresentationToModelIntents;
import com.better.alarm.model.AlarmSetter;
import com.better.alarm.model.AlarmValue;
import com.better.alarm.model.CalendarType;
import com.better.alarm.presenter.AlarmsListActivity;
import com.better.alarm.services.AlarmsService;
import com.google.common.base.Optional;

import org.hamcrest.Matcher;
import org.junit.Rule;
import org.junit.Test;
import org.junit.rules.RuleChain;
import org.junit.rules.TestRule;
import org.junit.runner.RunWith;

import java.util.Locale;

import cortado.Cortado;
import io.reactivex.annotations.NonNull;
import io.reactivex.functions.Predicate;

import static android.support.test.espresso.Espresso.onData;
import static android.support.test.espresso.Espresso.onView;
import static android.support.test.espresso.action.ViewActions.click;
import static android.support.test.espresso.assertion.ViewAssertions.matches;
import static android.support.test.espresso.matcher.ViewMatchers.isDisplayed;
import static android.support.test.espresso.matcher.ViewMatchers.withId;
import static android.support.test.espresso.matcher.ViewMatchers.withText;
import static org.hamcrest.Matchers.anything;

@RunWith(AndroidJUnit4.class)
@LargeTest
public class ListTest extends BaseTest {
    public ActivityTestRule<AlarmsListActivity> listActivity = new ActivityTestRule<AlarmsListActivity>(
            AlarmsListActivity.class, false, /* autostart*/ true);

    @Rule
    public TestRule chain = RuleChain.outerRule(new ForceLocaleRule(Locale.US)).around(listActivity);

    @Test
    public void newAlarmShouldBeDisabledIfNotEdited() throws Exception {
        sleep();
        onView(withId(R.id.fab)).perform(click());
        sleep();
        onView(withText("Cancel")).perform(click());
        sleep();
        Cortado.onView().withText("OK").perform().click();
        sleep();

        assertThatList().items().hasSize(3);

        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(enabled())
                .items()
                .isEmpty();

        deleteAlarm(0);
        assertThatList().items().hasSize(2);
    }

    @Test
    public void newAlarmShouldBeEnabledIfEdited12() throws Exception {
        AlarmApplication.is24hoursFormatOverride = Optional.of(false);
        newAlarmShouldBeEnabledIfEdited();
    }

    @Test
    public void newAlarmShouldBeEnabledIfEdited24() throws Exception {
        AlarmApplication.is24hoursFormatOverride = Optional.of(true);
        newAlarmShouldBeEnabledIfEdited();
    }

    private void newAlarmShouldBeEnabledIfEdited() throws Exception {
        onView(withId(R.id.fab)).perform(click());
        sleep();
        Cortado.onView().withText("1").perform().click();
        Cortado.onView().withText("2").perform().click();
        Cortado.onView().withText("3").perform().click();
        Cortado.onView().withText("5").perform().click();

        onView(withText("AM"))
                .withFailureHandler(new FailureHandler() {
                    @Override
                    public void handle(Throwable error, Matcher<View> viewMatcher) {
                    }
                })
                .perform(click());

        sleep();
        onView(withText("OK")).perform(click());
        Cortado.onView().withText("OK").perform().click();
        sleep();

        assertThatList().items().hasSize(3);

        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(enabled())
                .items()
                .hasSize(1);

        deleteAlarm(0);
        assertThatList().items().hasSize(2);
    }

    @Test
    public void testDeleteNewAlarmInDetailsActivity() throws Exception {
        onView(withId(R.id.fab)).perform(click());
        sleep();
        onView(withText("Cancel")).perform(click());
        Cortado.onView().withText("OK").perform().click();
        sleep();

        assertThatList().items().hasSize(3);
        sleep();

        onData(anything()).atPosition(0).onChildView(withId(R.id.details_button_container)).perform(click());
        sleep();

        Cortado.onView().withId(R.id.set_alarm_menu_delete_alarm).perform().click();
        sleep();

        Cortado.onView().withText("OK").perform().click();
        sleep();

        assertThatList().items().hasSize(2);
    }

    @Test
    public void newAlarmShouldBeDisabledAfterDismiss() throws Exception {
        onView(withId(R.id.fab)).perform(click());
        sleep();
        Cortado.onView().withText("1").perform().click();
        Cortado.onView().withText("2").perform().click();
        Cortado.onView().withText("3").perform().click();
        Cortado.onView().withText("5").perform().click();

        onView(withText("AM"))
                .withFailureHandler(new FailureHandler() {
                    @Override
                    public void handle(Throwable error, Matcher<View> viewMatcher) {
                        //ignore fails - only use if View is found
                    }
                })
                .perform(click());

        sleep();
        onView(withText("OK")).perform(click());
        Cortado.onView().withText("OK").perform().click();
        sleep();

        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(enabled())
                .items()
                .hasSize(1);

        int id = ListAsserts.<AlarmValue>listObservable(R.id.list_fragment_list).firstOrError().blockingGet().getId();

        //simulate alarm fired
        Intent intent = new Intent(AlarmSetter.ACTION_FIRED);
        intent.putExtra(AlarmSetter.EXTRA_ID, id);
        intent.putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.NORMAL.name());
        listActivity.getActivity().sendBroadcast(intent);

        sleep();

        //simulate dismiss from the notification bar
        Intent dismiss = new Intent(PresentationToModelIntents.ACTION_REQUEST_DISMISS);
        dismiss.putExtra(AlarmSetter.EXTRA_ID, id);
        dismiss.setClass(listActivity.getActivity(), AlarmsService.class);
        listActivity.getActivity().startService(dismiss);

        sleep();

        //alarm must be disabled because there is no repeating
        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(enabled())
                .items()
                .isEmpty();

        deleteAlarm(0);
        assertThatList().items().hasSize(2);
    }

    @Test
    public void editAlarmALot() throws Exception {
        onView(withId(R.id.fab)).perform(click());
        sleep();
        assertTimerView("--:--");
        Cortado.onView().withText("1").perform().click();
        assertTimerView("--:-1");
        Cortado.onView().withText("2").perform().click();
        assertTimerView("--:12");
        Cortado.onView().withText("3").perform().click();
        assertTimerView("-1:23");
        Cortado.onView().withText("5").perform().click();
        assertTimerView("12:35");
        Cortado.onView().withId(R.id.delete).perform().click();
        assertTimerView("-1:23");
        Cortado.onView().withId(R.id.delete).perform().click();
        assertTimerView("--:12");
        Cortado.onView().withId(R.id.delete).perform().longClick();
        assertTimerView("--:--");
        sleep();
        onView(withText("Cancel")).perform(click());
        sleep();
        onView(withText("Cancel")).perform(click());
        assertThatList().items().hasSize(2);
    }

    @Test
    public void editRepeat() throws Exception {
        sleep();
        onView(withId(R.id.fab)).perform(click());
        sleep();
        onView(withText("Cancel")).perform(click());
        sleep();

        onView(withText("Repeat")).perform(click());
        onView(withText("Monday")).perform(click());
        onView(withText("Tuesday")).perform(click());
        onView(withText("Wednesday")).perform(click());
        onView(withText("Thursday")).perform(click());
        onView(withText("Friday")).perform(click());
        onView(withText("Saturday")).perform(click());
        onView(withText("Sunday")).perform(click());
        onView(withText("OK")).perform(click());

        Cortado.onView().withText("OK").perform().click();
        sleep();
        sleep();
        
        assertThatList().items().hasSize(3);

        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(enabled())
                .items()
                .hasSize(1);

        ListAsserts.<AlarmValue>assertThatList(R.id.list_fragment_list)
                .filter(new Predicate<AlarmValue>() {
                    @Override
                    public boolean test(@NonNull AlarmValue alarmValue) throws Exception {
                        return alarmValue.getDaysOfWeek().isRepeatSet();
                    }
                })
                .items()
                .hasSize(3);

        onView(withText("Every day")).check(matches(isDisplayed()));

        deleteAlarm(0);
        assertThatList().items().hasSize(2);
    }

}

File: app/src/androidTest/java/com/better/alarm/test/ListTest.kt
package com.better.alarm.test

import android.content.Intent
import androidx.test.espresso.Espresso.onData
import androidx.test.espresso.Espresso.onView
import androidx.test.espresso.action.ViewActions.click
import androidx.test.espresso.action.ViewActions.longClick
import androidx.test.espresso.assertion.ViewAssertions.matches
import androidx.test.espresso.matcher.ViewMatchers.isCompletelyDisplayed
import androidx.test.espresso.matcher.ViewMatchers.isDisplayed
import androidx.test.espresso.matcher.ViewMatchers.withId
import androidx.test.espresso.matcher.ViewMatchers.withText
import androidx.test.ext.junit.rules.ActivityScenarioRule
import androidx.test.ext.junit.runners.AndroidJUnit4
import com.better.alarm.R
import com.better.alarm.bootstrap.overrideIs24hoursFormatOverride
import com.better.alarm.data.CalendarType
import com.better.alarm.domain.AlarmSetter
import com.better.alarm.receivers.AlarmsReceiver
import com.better.alarm.receivers.PresentationToModelIntents
import com.better.alarm.test.TestSync.Companion.clickFab
import com.better.alarm.ui.main.AlarmsListActivity
import java.util.*
import kotlinx.coroutines.runBlocking
import org.assertj.core.api.Assertions.assertThat
import org.hamcrest.Matchers.allOf
import org.hamcrest.Matchers.anything
import org.junit.After
import org.junit.Before
import org.junit.Rule
import org.junit.Test
import org.junit.rules.RuleChain
import org.junit.rules.TestRule
import org.junit.runner.RunWith

@RunWith(AndroidJUnit4::class)
class ListTest {
  @JvmField var listActivity = ActivityScenarioRule(AlarmsListActivity::class.java)

  @JvmField
  @Rule
  var chain: TestRule = RuleChain.outerRule(ForceLocaleRule(Locale.US)).around(listActivity)

  @Before
  fun drop() {
    dropDatabase()
  }

  @After
  fun after() {
    dropDatabase()
  }

  @Test
  fun newAlarmShouldBeDisabledIfNotEdited() {
    clickFab()
    onView(withText("Cancel")).perform(click())
    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(3)

    assertThat(alarmsList().filter { it.isEnabled }).isEmpty()
  }

  @Test
  fun alarmsCanBeDeletedWithLongClick() {
    clickFab()
    onView(withText("Cancel")).perform(click())
    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(3)

    onData(anything())
        .onChildView(withId(R.id.details_button_container))
        .atPosition(0)
        .perform(longClick())
    onView(withText("Delete alarm")).click()
    onView(withText("OK")).click()

    assertThat(alarmsList()).hasSize(2)
  }

  @Test
  fun newAlarmShouldBeEnabledIfEdited12() {
    overrideIs24hoursFormatOverride(false)
    newAlarmShouldBeEnabledIfEdited()
  }

  @Test
  fun newAlarmShouldBeEnabledIfEdited24() {
    overrideIs24hoursFormatOverride(true)
    newAlarmShouldBeEnabledIfEdited()
  }

  private fun newAlarmShouldBeEnabledIfEdited() {
    clickFab()
    onView(withText("1")).perform(click())
    onView(withText("2")).perform(click())
    onView(withText("3")).perform(click())
    onView(withText("5")).perform(click())

    onView(withText("AM"))
        .withFailureHandler { _, _ ->
          // ignore fails - only use if View is found
        }
        .perform(click())

    onView(withText("OK")).perform(click())
    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(3)

    assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)
  }

  @Test
  fun testDeleteNewAlarmInDetailsActivity() {
    clickFab()
    onView(withText("Cancel")).perform(click())
    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(3)

    onData(anything())
        .atPosition(0)
        .onChildView(withId(R.id.details_button_container))
        .perform(click())

    onView(withId(R.id.set_alarm_menu_delete_alarm)).perform(click())

    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(2)
  }

  @Test
  fun newAlarmShouldBeDisabledAfterDismiss() =
      runBlocking<Unit> {
        clickFab()
        onView(withText("1")).perform(click())
        onView(withText("2")).perform(click())
        onView(withText("3")).perform(click())
        onView(withText("5")).perform(click())

        onView(withText("AM"))
            .withFailureHandler { _, _ ->
              // ignore fails - only use if View is found
            }
            .perform(click())

        onView(withText("OK")).perform(click())
        onView(withText("OK")).perform(click())

        assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)

        val id = alarmsList().first { it.isEnabled }.id

        fireAlarm(id)
        dismissAlarm(id)

        // listActivity.scenario.onActivity { it.store.uiVisible.filter { it }.blockingFirst() }
        println("Dismissed (done)")

        // alarm must be disabled because there is no repeating
        assertThat(alarmsList().filter { it.isEnabled }).isEmpty()
      }

  @Test
  fun editAlarmALot() {
    clickFab()
    assertTimerView("--:--")
    onView(withText("1")).perform(click())
    assertTimerView("--:-1")
    onView(withText("2")).perform(click())
    assertTimerView("--:12")
    onView(withText("3")).perform(click())
    assertTimerView("-1:23")
    onView(withText("5")).perform(click())
    assertTimerView("12:35")
    onView(withId(R.id.delete)).perform(click())
    assertTimerView("-1:23")
    onView(withId(R.id.delete)).perform(click())
    assertTimerView("--:12")
    onView(withId(R.id.delete)).perform(longClick())
    assertTimerView("--:--")
    onView(withText("Cancel")).perform(click())
    onView(withText("Cancel")).perform(click())
    assertThat(alarmsList()).hasSize(2)
  }

  @Test
  fun editRepeat() {
    clickFab()
    onView(withText("Cancel")).perform(click())

    onView(withText("Repeat")).perform(click())

    onView(allOf(isCompletelyDisplayed(), withText("Monday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Tuesday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Wednesday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Thursday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Friday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Saturday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("Sunday"))).perform(click())
    onView(allOf(isCompletelyDisplayed(), withText("OK"))).perform(click())

    onView(withText("OK")).perform(click())

    assertThat(alarmsList()).hasSize(3)

    assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)

    assertThat(alarmsList().filter { it.isRepeatSet }).hasSize(3)

    onView(withText("Every day")).check(matches(isDisplayed()))
  }

  @Test
  fun changeTimeInList() {
    onData(anything()).atPosition(0).onChildView(withId(R.id.digital_clock_time)).perform(click())

    onView(withText("1")).perform(click())
    onView(withText("2")).perform(click())
    onView(withText("3")).perform(click())
    onView(withText("5")).perform(click())

    onView(withText("AM"))
        .withFailureHandler { error, viewMatcher ->
          // ignore fails - only use if View is found
        }
        .perform(click())

    onView(withText("OK")).perform(click())

    assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)

    onData(anything())
        .atPosition(0)
        .onChildView(withId(R.id.list_row_on_off_checkbox_container))
        .perform(click())
    assertThat(alarmsList().filter { it.isEnabled }).isEmpty()
  }

  /**
   * This is a test for https://github.com/yuriykulikov/AlarmClock/issues/361
   *
   * ## Given
   * * An alarm with Delete after dismissed activated
   *
   * ## When
   * * Alarm is fired
   * * Alarm is dismissed ## Then
   * * It should be deleted
   */
  @Test
  fun deleteOnDismiss() =
      runBlocking<Unit> {
        clickFab()
        onView(withText("1")).perform(click())
        onView(withText("2")).perform(click())
        onView(withText("3")).perform(click())
        onView(withText("5")).perform(click())

        onView(withText("AM"))
            .withFailureHandler { error, viewMatcher ->
              // ignore fails - only use if View is found
            }
            .perform(click())
        onView(withText("OK")).perform(click())
        onView(withText("OK")).perform(click())

        assertThat(alarmsList()).hasSize(3)
        assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)
        assertThat(alarmsList().first { it.isEnabled }.isDeleteAfterDismiss).isTrue

        // when fired and dismissed
        val id = alarmsList().first { it.isEnabled }.id
        fireAlarm(id)
        dismissAlarm(id)

        // then alarm must be deleted from the list
        assertThat(alarmsList()).hasSize(2)
      }

  /**
   * This is a test for https://github.com/yuriykulikov/AlarmClock/issues/361
   *
   * ## Given
   * * An alarm with Delete after dismissed activated
   *
   * ## When
   * * Alarm is fired
   * * Alarm is dismissed ## Then
   * * It should be deleted
   */
  @Test
  fun deleteOnDismissDeactivated() =
      runBlocking<Unit> {
        clickFab()
        onView(withText("Cancel")).perform(click())
        onView(withText("Delete after dismissed")).perform(click())
        onView(withText("OK")).perform(click())

        assertThat(alarmsList()).hasSize(3)
        assertThat(alarmsList().filter { it.isEnabled }).hasSize(1)
        assertThat(alarmsList().first { it.isEnabled }.isDeleteAfterDismiss).isFalse

        // when fired and dismissed
        val id = alarmsList().first { it.isEnabled }.id
        fireAlarm(id)
        dismissAlarm(id)

        // then alarm must still be present in the list
        assertThat(alarmsList()).hasSize(3)
      }

  private fun fireAlarm(id: Int) {
    listActivity.scenario.onActivity { activity ->
      activity.sendBroadcast(
          Intent().apply {
            action = AlarmSetter.ACTION_FIRED
            setClass(activity, AlarmsReceiver::class.java)
            putExtra(AlarmSetter.EXTRA_ID, id)
            putExtra(AlarmSetter.EXTRA_TYPE, CalendarType.NORMAL.name)
          })
    }
  }

  private suspend fun dismissAlarm(id: Int) {
    sentTestIntent { context ->
      action = PresentationToModelIntents.ACTION_REQUEST_DISMISS
      setClass(context, AlarmsReceiver::class.java)
      putExtra(AlarmSetter.EXTRA_ID, id)
    }
  }
}
